Data/Unicode/Internal/NormalizeStream.hs

```diff
@@ -211,124 +211,135 @@ maxDecomposeLen = 32
 -- Composition
 -------------------------------------------------------------------------------
 
-composeAndWrite
-    :: A.MArray s
-    -> Int
-    -> Char
-    -> ReBuf
-    -> ST s Int -- return new index
-composeAndWrite arr di st Empty = do
-    n <- unsafeWrite arr di st
-    return (di + n)
-
-composeAndWrite arr di st (One c) =
-    case C.composePair st c of
-        Just x -> do
-            n <- unsafeWrite arr di x
-            return (di + n)
-        Nothing -> do
-            n <- unsafeWrite arr di st
-            m <- unsafeWrite arr (di + n) c
-            return (di + n + m)
-
-composeAndWrite arr di st (Many c1 c2 str) =
-    composeAndWrite' arr di st (c1 : c2 : str)
-
-composeAndWrite'
-    :: A.MArray s
-    -> Int
-    -> Char
-    -> [Char]
-    -> ST s Int
-composeAndWrite' arr di = go []
-    where
-        -- arguments: uncombined chars, starter, unprocessed str
-        go uncs st [] = writeStr arr di (st : uncs)
-        go uncs st (c : cs) = case C.composePair st c of
-            Nothing -> go (uncs ++ (c : same)) st bigger
-            Just x  -> go uncs x cs
-            where
-                cc = CC.getCombiningClass c
-                (same, bigger) = span ((== cc) . CC.getCombiningClass) cs
-
 -------------------------------------------------------------------------------
 -- Composition of Hangul Jamo characters, done algorithmically
 -------------------------------------------------------------------------------
 
+-- If we are composing we do not need to first decompose Hangul. We can just
+-- compose assuming there could be some partially composed syllables e.g. LV
+-- syllable followed by a jamo T. We need to compose this case as well.
+--
+-- XXX The unicode normalization test suite does not seem to have tests for a
+-- LV composed hangul syllable followed by a jamo T.
+
 -- Hold an L to wait for V, hold an LV to wait for T.
 data JamoBuf
-    = JamoEmpty
-    | JamoLIndex {-# UNPACK #-} !Int
+    = JamoLIndex {-# UNPACK #-} !Int
```

Comment:
I believe `{-# UNPACK #-}` pragmas can be safely omitted, there is `{-# OPTIONS_GHC -funbox-strict-fields #-}` anyways.

Reply:
yes, it can be removed.